File: mobile/ios/SpeakeasyCrypto/SignalStoreAdapter.swift
import Foundation
import SignalClient

// Adapter that bridges our local SpeakeasyStore (Keychain) to LibSignal's Requirements

class SignalStoreAdapter: IdentityKeyStore, PreKeyStore, SessionStore, SignedPreKeyStore {
    
    private let store: SpeakeasyStore
    
    init(store: SpeakeasyStore) {
        self.store = store
    }
    
    // MARK: - IdentityKeyStore
    func getIdentityKeyPair() -> IdentityKeyPair {
        guard let data = store.getIdentityKeyPair() else {
            fatalError("Identity Key not generated yet!")
        }
        return try! IdentityKeyPair(from: data)
    }
    
    func getLocalRegistrationId() -> UInt32 {
        guard let id = store.getLocalRegistrationId() else {
            fatalError("Registration ID not generated yet!")
        }
        return id
    }
    
    func saveIdentity(_ identity: IdentityKeyPair?, registrationId: UInt32?) {
        // We usually generate initally, but this fulfills the protocol if needed.
        if let i = identity, let r = registrationId {
            store.saveIdentity(i.serialize(), registrationId: r)
        }
    }
    
    func isTrustedIdentity(_ identity: SignalClient.IdentityKey, isolationGroup: IsolationGroup?, direction: SignalClient.Direction, address: SignalClient.ProtocolAddress) -> Bool {
        // Trust On First Use (TOFU) or Always Trust for Phase 2
        return true
    }
    
    func getIdentity(address: SignalClient.ProtocolAddress) -> SignalClient.IdentityKey? {
        // We don't persist remote identities in Phase 2 yet (Trust all)
        return nil 
    }
    
    // MARK: - PreKeyStore
    func loadPreKey(id: UInt32) throws -> PreKeyRecord {
        guard let data = store.loadPreKey(id: id) else {
            throw SignalError.invalidKey
        }
        return try PreKeyRecord(from: data)
    }
    
    func storePreKey(_ record: PreKeyRecord, id: UInt32) throws {
        store.storePreKey(record.serialize(), id: id)
    }
    
    func containsPreKey(id: UInt32) -> Bool {
        return store.containsPreKey(id: id)
    }
    
    func removePreKey(id: UInt32) throws {
        store.removePreKey(id: id)
    }
    
    // MARK: - SignedPreKeyStore
    func loadSignedPreKey(id: UInt32) throws -> SignedPreKeyRecord {
        guard let data = store.loadSignedPreKey(id: id) else {
            throw SignalError.invalidKey
        }
        return try SignedPreKeyRecord(from: data)
    }
    
    func storeSignedPreKey(_ record: SignedPreKeyRecord, id: UInt32) throws {
        store.storeSignedPreKey(record.serialize(), id: id)
    }
    
    func containsSignedPreKey(id: UInt32) -> Bool {
        return store.containsSignedPreKey(id: id)
    }
    
    func removeSignedPreKey(id: UInt32) throws {
        store.removeSignedPreKey(id: id)
    }
    
    // MARK: - SessionStore
    func loadSession(address: SignalClient.ProtocolAddress) -> SessionRecord? {
        if let data = store.loadSession(address: address.name) {
             return try? SessionRecord(from: data)
        }
        return nil // Return nil or empty record depending on libsignal version
    }
    
    func storeSession(_ record: SessionRecord, address: SignalClient.ProtocolAddress) throws {
        store.storeSession(record.serialize(), address: address.name)
    }
    
    func containsSession(address: SignalClient.ProtocolAddress) -> Bool {
        return store.containsSession(address: address.name)
    }
    
    func deleteSession(address: SignalClient.ProtocolAddress) throws {
        store.deleteSession(address: address.name)
    }
}

File: mobile/android/crypto/SignalStoreAdapter.kt
package com.speakeasy.crypto

import org.signal.libsignal.IdentityKey
import org.signal.libsignal.IdentityKeyPair
import org.signal.libsignal.SignalProtocolAddress
import org.signal.libsignal.state.PreKeyRecord
import org.signal.libsignal.state.SessionRecord
import org.signal.libsignal.state.SignalProtocolStore
import org.signal.libsignal.state.SignedPreKeyRecord
import java.io.IOException

class SignalStoreAdapter(private val store: SpeakeasyStore) : SignalProtocolStore {

    // IdentityKeyStore
    override fun getIdentityKeyPair(): IdentityKeyPair {
        val data = store.getIdentityKeyPair() ?: throw RuntimeException("No Identity Key")
        return IdentityKeyPair(data)
    }

    override fun getLocalRegistrationId(): Int {
        return store.getLocalRegistrationId() ?: throw RuntimeException("No Registration ID")
    }

    override fun saveIdentity(address: SignalProtocolAddress, identityKey: IdentityKey): Boolean {
        // TOFU: Always trust for now
        return true
    }

    override fun isTrustedIdentity(address: SignalProtocolAddress, identityKey: IdentityKey, direction: IdentityKeyStore.Direction): Boolean {
        return true
    }

    override fun getIdentity(address: SignalProtocolAddress): IdentityKey? {
        // We don't persist remote identities yet
        return null
    }

    // PreKeyStore
    override fun loadPreKey(preKeyId: Int): PreKeyRecord {
        val data = store.loadPreKey(preKeyId) ?: throw org.signal.libsignal.InvalidKeyIdException("PreKey not found")
        return PreKeyRecord(data)
    }

    override fun storePreKey(preKeyId: Int, record: PreKeyRecord) {
        store.storePreKey(record.serialize(), preKeyId)
    }

    override fun containsPreKey(preKeyId: Int): Boolean {
        return store.containsPreKey(preKeyId)
    }

    override fun removePreKey(preKeyId: Int) {
        store.removePreKey(preKeyId)
    }

    // SignedPreKeyStore
    override fun loadSignedPreKey(signedPreKeyId: Int): SignedPreKeyRecord {
        val data = store.loadSignedPreKey(signedPreKeyId) ?: throw org.signal.libsignal.InvalidKeyIdException("SignedPreKey not found")
        return SignedPreKeyRecord(data)
    }

    override fun loadSignedPreKeys(): List<SignedPreKeyRecord> {
        return emptyList() // Not used in basic flow
    }

    override fun storeSignedPreKey(signedPreKeyId: Int, record: SignedPreKeyRecord) {
        store.storeSignedPreKey(record.serialize(), signedPreKeyId)
    }

    override fun containsSignedPreKey(signedPreKeyId: Int): Boolean {
        return store.containsSignedPreKey(signedPreKeyId)
    }

    override fun removeSignedPreKey(signedPreKeyId: Int) {
        store.removeSignedPreKey(signedPreKeyId)
    }

    // SessionStore
    override fun loadSession(address: SignalProtocolAddress): SessionRecord {
        val data = store.loadSession(address.toString())
        return if (data != null) SessionRecord(data) else SessionRecord()
    }

    override fun getSubDeviceSessions(name: String): List<Int> {
        return emptyList() 
    }

    override fun storeSession(address: SignalProtocolAddress, record: SessionRecord) {
        store.storeSession(record.serialize(), address.toString())
    }

    override fun containsSession(address: SignalProtocolAddress): Boolean {
        return store.containsSession(address.toString())
    }

    override fun deleteSession(address: SignalProtocolAddress) {
        store.deleteSession(address.toString())
    }

    override fun deleteAllSessions(name: String) {
        // No-op
    }
}
